Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/NonCompositedFrameRenderer.cpp

```diff
@@ -87,16 +112,50 @@ void NonCompositedFrameRenderer::display()
     auto* graphicsContext = m_surface->graphicsContext();
     if (!graphicsContext || !m_context->makeContextCurrent())
         return;
+#if ENABLE(DAMAGE_TRACKING)
+    if (m_frameDamage) {
+        {
+            Locker locker { m_frameDamageHistoryForTestingLock };
+            if (m_frameDamageHistoryForTesting)
+                m_frameDamageHistoryForTesting->append(m_frameDamage->regionForTesting());
+        }
+        m_surface->setFrameDamage(WTF::move(*m_frameDamage));
+        resetFrameDamage();
+    }
+#endif
 
-    // FIXME: Use render target damage not to re-paint whole rect.
-    webPage->drawRect(*graphicsContext, webPage->bounds());
+    auto rectToRepaint = webPage->bounds();
+#if ENABLE(DAMAGE_TRACKING)
+    if (webPage->corePage()->settings().useDamagingInformationForCompositing()) {
+        if (auto& renderTargetDamage = m_surface->renderTargetDamage())
+            rectToRepaint = renderTargetDamage->bounds();
+    }
+#endif
+    webPage->drawRect(*graphicsContext, rectToRepaint);
```

Comment:
I'm not sure we should make this conditional. Propagating damage to the UI process depends on damage tracking, but in non-composited mode we have always rendered only the damage region, so I think we should do the same here. Can't we just use the dirty region here instead of the damage from the surface?